### ISSUE TITLE
Replace notable interfaces with abstract classes, for multiple reasons

### ISSUE THREAD

**Issue by airbreather** [NetTopologySuite/GeoAPI#68]
As a sort-of alternative to #67, if we replace interfaces like `IGeometry` with abstract classes like `GeometryBase`, then we can provide a much better experience for implementers while still retaining the same essential API shape (without resorting to "tricks" to make almost-"default interface methods" happen).

I expect that this will be somewhat more work than #67 to get up and running, but I also claim that it is likely to be a **much** better experience for all parties (except maybe the OGC folks, who would likely be indifferent because it's not really in the scope of what they care about).

These are the interfaces that I propose replacing with abstract classes:
# Common Themes
1. `FooCore` methods are protected members that subclasses may (or must) override which are used to implement one or more non-virtual `Foo` members.
    - These exist instead of making `Foo` itself virtual / abstract, because there's often surface-level parameter validation / short-circuiting that all implementations of `Foo` can legally do the same way, so why force every implementer to repeat it?
1. I would like to figure out the "right" way to handle `Ordinate` / `Ordinates` / `Dimension` / `Measures`.
    - **For now**, I'll pretend like we've resolved it in a way that defines `GetOrdinate(i, Ordinate.Z)` as **always** being identical to `GetZ(i)`; if we resolve it differently, then some things will have to adjust.

# `IPreparedGeometry`
This would be a subset of NTS's `BasicPreparedGeometry`.  Accepts a geometry instance in the constructor that it uses, by default for all the operations.  And that's it.

No abstract members; an empty subclass would be completely functional (but uninteresting).

# `ICoordinateSequence`
## Constructors
1. `protected CoordinateSequenceBase(int count, Ordinates includedOrdinates)`
    - `count >= 0`
    - :thinking: validate that `includedOrdinates` has at least the`Ordinates.XY` flags? 
    - :thinking: validate that `includedOrdinates` has at most the `Ordinates.XYZM` flags?
1. `protected CoordinateSequenceBase(CoordinateSequenceBase copyFrom)`
    - Invoke the other constructor using the values of `copyFrom`

## Abstract methods
1. `double GetOrdinateCore(int i, Ordinate ordinate)`
1. `void SetOrdinateCore(int i, Ordinate ordinate, double value)`
1. `CoordinateSequenceBase Copy();`

## Virtual methods
1. `Coordinate CreateCoordinate()`
    - Default is to create `Coordinate`, `CoordinateZ`, `CoordinateM`, or `CoordinateZM` as appropriate for our ordinate mask, otherwise throw `NotSupportedException`.
        - `NotSupportedException` path could be skipped if we do both :thinking: constructor validations.
1. `Envelope ExpandEnvelope(Envelope env)`
    - Default does the obvious loop.
    - Subclass might want control of the loop for performance reasons.
1. `Coordinate GetCoordinateCore(int i)`
    - Default calls `CreateCoordinate` and initializes as appropriate.
    - Subclass might want to override to avoid allocating and copying.
1. `Coordinate GetCoordinateCopyCore(int i)`
    - Default calls `CreateCoordinate` and initializes as appropriate.
    - Subclass might want to override to avoid allocating and copying.
1. `Coordinate[] ToCoordinateArray()`
    - Default creates the array the obvious way
    - Subclass might want to override to avoid allocating and copying.
1. `void CopyFromCore(CoordinateSequenceBase other, int offset, int count)`
    - Default copies ordinates.
    - Subclass will **very likely** want to override for performance reasons to avoid virtual call overhead and (perhaps) to use direct array copying if the source stores its data in a compatible format.

## Other discussion
Other methods are non-virtual and implemented only in terms of what we have available above.

# `ICoordinateSequenceFactory`
## Constructors
1. `protected CoordinateSequenceFactoryBase(Ordinates maxOrdinates)`
    - :thinking: validate that `includedOrdinates` has at least the`Ordinates.XY` flags? 
    - :thinking: validate that `includedOrdinates` has at most the `Ordinates.XYZM` flags?

## Abstract methods
1. `CoordinateSequenceBase CreateCore(int count, Ordinates ordinates)`

## Virtual methods
1. `CoordinateSequenceBase CreateCore(CoordinateSequenceBase copyFrom, Ordinates ordinates)`
    - Default does something like `result = CreateCore(copyFrom.Count, ordinates)`, then `result.CopyFrom(copyFrom)`, then returns it.
    - Subclass may want to override if initializing an empty sequence involves significant work that would have to get repeated anyway during `CopyFrom`.

## Other discussion
If we want to keep a `Create` method(s) that accept(s) `Coordinate[]`, should `CoordinateArraySequence` move up to `GeoAPI`?  It could help a lot, both here and elsewhere...

# `IGeometry`
## Constructors
1. `protected GeometryBase(GeometryFactoryBase factory, OgcGeometryType ogcGeometryType)`

## Abstract members
This still seems like a little too much for implementers to have to implement, but I think we risk going too far the other way if we cut out too much more.

1. `GeometryBase Boundary { get; }`
1. `Dimension BoundaryDimension { get; }`
1. `GeometryBase ComputeCentroidCore()`
1. `GeometryBase ComputeInteriorPointCore()`
1. `void Normalize()`
1. `GeometryBase Copy()`
1. `GeometryBase ConvexHull()`
1. `IntersectionMatrix Relate(GeometryBase g)`
1. `GeometryBase Overlay(GeometryBase g, SpatialFunction opCode)`
    - Not in here yet, but it seems to belong here for the same reason that `Relate` does...
1. `GeometryBase BufferCore(double distance, BufferParameters bufferParameters)`
    - Technically we could keep `IBufferParameters`, but I doubt that there's any sort of need for a different implementation than the one that's in NTS...
1. `bool EqualsExact(GeometryBase other, double tolerance)`
1. `double Distance(GeometryBase other)`
1. `void Apply(ICoordinateFilter filter)`
1. `void Apply(ICoordinateSequenceFilter filter)`
1. `void Apply(IGeometryFilter filter)`
1. `void Apply(IGeometryComponentFilter filter)`
1. `void GeometryChanged()`
1. `byte[] AsBinary()`
1. `string AsText()`

## Virtual members
1. `double Area { get; }`
    - Default is 0
1. `double Length { get; }`
    - Default is 0
1. `int NumGeometries { get; }`
    - Default runs `Apply(IGeometryFilter)`, returns how many times `Filter` was called.
1. `int NumPoints { get; }`
    - Default runs `Apply(ICoordinateSequenceFilter)`, returns how many times `Filter` was called.
1. `Coordinate Coordinate { get; }`
    - Default runs `Apply(ICoordinateFilter)` and returns the first one.
1. `Coordinate[] Coordinates { get; }`
    - Default runs `Apply(ICoordinateFilter)` and returns all of them.
1. `double[] GetOrdinates(Ordinate ordinate)`
    - Default runs `Apply(ICoordinateSequenceFilter)` and returns the values.
1. `Dimension Dimension { get; }`
    - Default checks the `OgcGeometryType`.
        - All defined values other than `GeometryCollection` have a corresponding result.
        - For `GeometryCollection`, run `Apply(IGeometryFilter)` and return the max it sees (or False).
        - Undefined values throw `NotSupportedException`.
1. `Envelope EnvelopeInternal { get; }`
    - Default runs `Apply(ICoordinateSequenceFilter)`, returns an `Envelope` that contains all.
1. `GeometryBase GetGeometryN(int n)`
    - Default is itself for 0, otherwise runs `Apply(IGeometryFilter)` and returns the `n`th value.
1. Methods that NTS implements ultimately via `Relate`
    - Default is to **just** use `Relate`.
    - Special mention, only have one of the methods in each of the following pairs be virtual:
        - `Contains` / `Within`
        - `Covers` / `CoveredBy`
        - `Intersects` / `Disjoint`
1. Methods that NTS implements ultimately via `SnapIfNeededOverlayOp`
    - Default is to **just** use `Overlay`.
1. `bool IsEmpty { get; }`
    - Default is to run `Apply(ICoordinateSequenceFilter)` and return a value based on whether or not `Filter` was ever called.
1. `bool IsWithinDistance(GeometryBase geom, double distance)`
    - Default is to compare `distance` to the result of `Distance(geom)`
1. `void GeometryChangedAction()`
    - Default is to do nothing.

## Other discussion
`Union()` (the parameterless overload) is of questionable benefit at this layer.  It's not OGC, and if everything's valid, then the result is topologically identical to the original one.  Plus, it's semantically equivalent to `Union(EMPTY)`.  This isn't to say it's not useful, just that it doesn't seem useful **here**.

If we're going to have things like `Area` and `Centroid` on the base (which I actually think makes a ton of sense given that we don't have multiple inheritance), then we should probably also bring up the rest of the OGC methods still sitting on derived interfaces:
1. `ICurve.StartPoint` / `EndPoint`
1. `ICurve.IsClosed`
1. `ILineString.PointN`
1. `IPoint.X` / `Y` /  `Z` / `M`
1. `IPolygon.ExteriorRing` / `NumInteriorRings` / `GetInteriorRingN`

# TODO:
## `IPrecisionModel`

**Comment by FObermaier** [comment]
I object to any changes related to 
* ICoordinateSequence
* ICoordinateSequneceFactory

as these are core interfaces in JTS. Thus it puts an additional effort to port changes made in NTS.

If we lay hand on redesigning GeoAPI I'm in strong favor of removing all `IGeometry` interfaces or and instead come up with a set of interfaces / base classes that 
* just define the components that make up the geometry
* represent the hierarchy in OGC SFS 
* have few members which are easy to implement.

As a rule of thumb these new interfaces should reuse old names (OGC SFS) to make the transition with NTS easier.
 
These lean objects could be used with all NTS operation/algorithm classes independently, without having a full fledged `NetTopologySuite.Geometries.Geometry` object.

This answer applies largely to #67 as well.

**Comment by airbreather** [comment]
Sorry for the long comment, I don't have the time to write a short one and I wanted to address all your points without sounding dismissive or combative.

> I object to any changes related to
> 
> * ICoordinateSequence
> * ICoordinateSequneceFactory
> 
> 
> as these are core interfaces in JTS. Thus it puts an additional effort to port changes made in NTS.

I disagree on this point, specifically.

For one, Java has default interface methods, and JTS has recently decided to start actually using them in exactly these interfaces.  This feature is a long way off from being something we can use in GeoAPI*, and so using abstract classes would actually make these things easier to port, not harder, **especially** if this happens again post-v2 and we aren't comfortable with taking a breaking change.
* *unless we drop support for the "classic" .NET Framework, which seems like a mistake and would still likely put it months away

I do understand and agree that there is an overall risk that JTS does something with one of its interfaces that we can't do because we lack multiple inheritance.  I think this risk is low, but if it does happen, then it would make porting difficult.

> If we lay hand on redesigning GeoAPI I'm in strong favor of removing all `IGeometry` interfaces or and instead come up with a set of interfaces / base classes that
> 
> * just define the components that make up the geometry
> * represent the hierarchy in OGC SFS
> * have few members which are easy to implement.

I totally agree that this would be great *if* we could pull it off without sacrificing usability too much for downstream folks like EF Core's users (as **an example**) who care a lot more about what you can do with geometry instances than about what they look like.  I question our ability to pull that off.

This feels unacceptable...:
```csharp
var queryLoc = geometryFactory.CreatePoint(new Coordinate(userLon, userLat));
var distance = MilesToMeters(20);
var gasStations = from gasStation in GetGasStations()
                  where DistanceOp.Distance(queryLoc, gasStation.Location) < distance
                  select gasStation;
```

...because if that query runs server-side, then so too should this equivalent formulation...:
```csharp
var queryLoc = geometryFactory.CreatePoint(new Coordinate(userLon, userLat));
var distance = MilesToMeters(20);
var gasStations = from gasStation in GetGasStations()
                  where new DistanceOp(queryLoc, gasStation.Location).Distance() < distance
                  select gasStation;
```

...and even this one...:
```csharp
var queryLoc = geometryFactory.CreatePoint(new Coordinate(userLon, userLat));
var distance = MilesToMeters(20);
var gasStations = from gasStation in GetGasStations()
                  let op = new DistanceOp(queryLoc, gasStation.Location)
                  where op.Distance() < distance
                  select gasStation;
```

...and so on.  Having basic operations as instance methods on the objects themselves feels like a requirement.  Even extension methods wouldn't quite replace it, because we don't have "extension properties" for things that are rightly implemented as properties.

What do you have in mind with such a redesign?  **Just** to put down something more concrete, here's an idea I've been toying with for a while (spoiler: I don't like it, and I'll get into why I don't later):
```csharp
public interface IGeometry
{
    /* a tiny set of members */
    /* can be fully implemented without complicated math */
}

public interface ISurface : IGeometry { }

public interface IPolygonal { }

public interface IPolygon : ISurface, IPolygonal
{
    /* extra polygon-specific stuff */
}

/* and so on for the other types */

public interface IUsefulGeometry
{
    IGeometry UnderlyingGeometry { get; }

    double Area { get; }
    double Length { get; }

    bool Intersects(IGeometry other);
    bool Covers(IGeometry other);
    IGeometry Union(IGeometry other);
    IGeometry Intersection(IGeometry other);

    /* etc., this is pretty much all the mathematically interesting parts */
}
```
Then the NTS implementations would be:
```csharp
public abstract class Geometry : IGeometry, IUsefulGeometry
{
    IGeometry IUsefulGeometry.UnderlyingGeometry => this;

    /* the rest are omitted, NTS does this */
}

public class Polygon : Geometry, IPolygon
{
    /* omitted, NTS does this */
}

/* and so on for the other types */
```
but someone would be free to do this without having to implement the `IGeometry` parts:
```csharp
public sealed class AlternativeGeometry : IUsefulGeometry
{
    public AlternativeGeometry(IGeometry underlying)
    {
        this.UnderlyingGeometry = underlying;
    }

    public IGeometry UnderlyingGeometry { get; }

    public bool Intersects(IGeometry other)
    {
        return AlternativeOp.Intersects(this.UnderlyingGeometry, other);
    }

    public IGeometry Union(IGeometry other)
    {
        return AlternativeOp.Union(this.UnderlyingGeometry, other);
    }

    /* etc. */
}
```

Is that something along the lines of what you're thinking of?

In fairness to the idea, I cannot deny that I love the fact that it adheres to the [SRP](https://en.wikipedia.org/wiki/Single_responsibility_principle), and it does seem like it could have the potential to expand the landscape of implementations.  I imagine that every implementer would usually have a somewhat easier time working with this, and in the worst case it's never outright **harder** because the combination of `IGeometry` + `IUsefulGeometry` more-or-less equates to the existing `IGeometry`.

However, I've been thinking about this on-and-off for a few years now, and I keep coming to the conclusion that it doesn't feel like the benefits to implementers would be worth the drawbacks that consumers would have to live with, where they could easily find themselves holding an `IGeometry` that they can't quite do anything with (except maybe plot on a map or save to a database or something).

This is especially discouraging because it's not (to my knowledge) how any of our counterparts in other languages work.  JTS, GEOS, and JSTS all bundle the concepts together.

At the end of the day, just being able to describe a geometry doesn't seem very useful in its own right (except to make it a bit easier to implement alternatives of NTS), and yet the only way I can see the idea working is to make sure to accept + return instances of that "description-only" object wherever we can.  Otherwise, `IUsefulGeometry` degenerates to a more complicated version of today's `IGeometry`.

---

I'd really love it if someone could convince me to change my mind on this, though, because I have some really cool ideas for how to optimize storing + saving + loading geometry instances that have very little to do with the math that you can do with them.

**Comment by FObermaier** [comment]
There would be no `IUsefulGeometry` interface just NTS `Geometry` classes implementing `IGeometryDefiniton` or whatever it is named.

EF would have to use NTS `Geometry`, afterall JTS doesn't have this set of `IGeometry` interfaces.
We could extent `GeometryFactory` to have some `Create(IGeometryDefinition gd)` overload.
That'd be it.

As I said, `IGeometryDefinition` would include all members of `IGeometry` that don't rely on any computational knowledge, like e.g. 

```C#
interface IGeometryDefinition {
   OgcGeometryType OgcGeometryType { get; }
   int SRID { get; }
   int NumPoints {get;}
   IEnumerable<Coordinate> Coordinates { get; }
   // there may be more
}

interface IPointDefinition : IGeometryDefinition {
   ICoordinateSequence CoordinateSequence { get; }
}

interface ILineStringDefinition : IGeometryDefinition, ILineal {
   ICoordinateSequence CoordinateSequence { get; }
}

interface IPolygonDefinition : IGeometryDefinition, IPolygonal {
   ILineStringDefinition ExteriorRing { get; }
   int NumInteriorRings { get; }
   ILineStringDefinition GetInteriorRingN(int index);
   //or
   //IList<ILineStringDefinition> InteriorRings { get; }
   //or
   //IEnumerable<ILineStringDefinition> InteriorRings { get; }
}

// ... to be continued
```

**Comment by airbreather** [comment]
I'm intrigued, but this stands out to me:

> EF would have to use NTS `Geometry`

My concern, which I attempted to telegraph in my choice of the name `IUsefulGeometry`, is that it wouldn't just be EF: in practice, I feel nearly **everybody** would have to use NTS `Geometry` (and with today's field of implementations, I'm certain that it's already a distinction without a difference).  `IGeometryDefinition` on its own just doesn't seem useful enough to justify its existence independent of NTS `Geometry`.

What are some examples of solutions that would use this separate thing called `IGeometryDefinition` without also needing the services that NTS `Geometry` provides?
- I guess I/O modules would be the best fit because this would open the door for them to reasonably define their own implementations.  But...
    - There are not terribly many of those (even fewer that we haven't implemented already)
    - Injecting `IGeometryFactory` and/or `IGeometryServices` seems more than adequate
    - "Extra" services available by downcasting isn't leaps and bounds better than `IFeature`.
- SharpMap seems like it may be a candidate (I haven't looked at it in much detail).  But...
    - Doesn't even that have to clip / ensure valid topology / do basic intersection testing?

I can't think of other significant problems that can be solved by `IGeometry` which would only actually need things on `IGeometryDefinition`.  Even if there are some others I'm missing, we also have a responsibility to weigh the benefits to them against the costs to everyone else who would have to switch their GeoAPI solution to use NTS, plus the costs to us for continuing to maintain this abstraction.

## In other words...
In other words, if we're OK with just splitting out this `IGeometryDefinition` piece and directing people to NTS (or similar alternative library) if they want to do anything interesting with the geometry instances, what are the reasons not to take it a step further and **remove GeoAPI.NET altogether**, with its remaining useful types going into NTS / ProjNet4GeoAPI as appropriate?

## Why do we even bother today?
My understanding was that the reason why we continue to maintain GeoAPI.NET (besides backwards-compatibility) is because a possible alternative implementation could implement the common operations (overlay, relate, buffer, distance, etc.) for higher-dimensional objects:
- "how far away is this airplane in-flight from the end of that runway?" and "how far away is this gas station from my car?" are different questions that can be modeled the same way:
    ```csharp
    IPoint point1 = GetPoint1(); // airplane or gas station
    IPoint point2 = GetPoint2(); // runway or my car

    // NTS can do gas station / car because it's on the surface
    // some other library would consider the elevation of the airplane
    // both use the common GeoAPI to describe "the same" request
    double distance = point1.Distance(point2);
    ```
- "I want to represent the zone that contains everything in a 20-meter sphere around either of my two satellites" and "I want to represent the zone that contains everything within 20 meters either my house or my work" are also identical:
    ```csharp
    IPoint point1 = GetPoint1(); // first satellite or my house
    IPoint point2 = GetPoint2(); // second satellite or my work

    // NTS can do the house / work after doing NTS issue #233
    // some other library would make the spherical thingy for the satellites
    // again, both use GeoAPI to make "the same" request
    IGeometry zone = point1.Buffer(20).Union(point2.Buffer(20));
    ```

## It's not free
If that's not actually what's going on, and furthermore some of the most useful bits of `IGeometry` can disappear from GeoAPI with no substitute (in GeoAPI), then I'm missing the value that justifies continuing the trend of maintaining separate "GeoAPI" versions of these into the vNext series of NTS / ProjNet4GeoAPI:
- Git repository
- Issue tracker
- Visual Studio solution
- Namespaces
- NuGet packages

**Comment by airbreather** [comment]
As suggested in https://github.com/NetTopologySuite/GeoAPI/issues/70#issuecomment-478899698, now that `NetTopologySuite.Core` has absorbed everything from `GeoAPI.Core`, I'm removing the following interfaces, limiting this first round to just the ones that should be uncontroversial because JTS uses classes instead (all in the `NetTopologySuite.Geometries` namespace):

1. `ICurve`
1. `IGeometry`
1. `IGeometryCollection`
1. `IGeometryFactory`
1. `ILinearRing`
1. `ILineString`
1. `IMultiCurve`
1. `IMultiLineString`
1. `IMultiPoint`
1. `IMultiPolygon`
1. `IMultiSurface`
1. `IPoint`
1. `IPolygon`
1. `IPrecisionModel`
1. `ISurface`

This subset of remaining interfaces in `NetTopologySuite.Core`, I am content with keeping as interfaces:
1. `NetTopologySuite.Algorithm.IBoundaryNodeRule`
1. `NetTopologySuite.Algorithm.IPointInAreaLocator`
1. `NetTopologySuite.Algorithm.Locate.IPointOnGeometryLocator`
1. `NetTopologySuite.Algorithm.Match.ISimilarityMeasure`
1. `NetTopologySuite.Geometries.IBoundable<out T, out TItem>`
1. `NetTopologySuite.Geometries.ICoordinateFilter`
1. `NetTopologySuite.Geometries.ICoordinateSequenceFilter`
1. `NetTopologySuite.Geometries.IExpandable<T>`
1. `NetTopologySuite.Geometries.IGeometryComponentFilter`
1. `NetTopologySuite.Geometries.IGeometryFilter`
1. `NetTopologySuite.Geometries.IIntersectable<in T>`
1. `NetTopologySuite.Geometries.ILineal`
1. `NetTopologySuite.Geometries.IPolygonal`
1. `NetTopologySuite.Geometries.IPuntal`
1. `NetTopologySuite.Geometries.Prepared.IPreparedGeometry`
1. `NetTopologySuite.Geometries.Utilities.GeometryEditor.IGeometryEditorOperation`
1. `NetTopologySuite.Geometries.Utilities.GeometryMapper.IMapOp`
1. `NetTopologySuite.Index.IIndexVisitor`
1. `NetTopologySuite.Index.IItemVisitor<in T>`
1. `NetTopologySuite.Index.ISpatialIndex<T>`
1. `NetTopologySuite.Index.KdTree.IKdNodeVisitor<T>`
1. `NetTopologySuite.Index.Strtree.IItemDistance<T, TItem>`
1. `NetTopologySuite.Index.Sweepline.ISweepLineOverlapAction`
1. All interfaces in the `NetTopologySuite.Noding` namespace
    - Because I don't understand the implementation well enough to be able to confidently propose an alternative.  `ISegmentString` in particular looks questionable.
1. `NetTopologySuite.Triangulate.ConstraintVertexFactory`
    - Though maybe it should be `IConstraintVertexFactory`
1. `NetTopologySuite.Triangulate.IConstraintSplitPointFinder`
1. `NetTopologySuite.Triangulate.QuadEdge.IQuadEdgeLocator`
1. `NetTopologySuite.Triangulate.QuadEdge.ITraversalVisitor`
1. `NetTopologySuite.Triangulate.QuadEdge.ITriangleVisitor`

That leaves the following ones, which I will claim that I do not think they should exist in their current form for one reason or another, in comments after this one:
1. `NetTopologySuite.IGeometryServices`
1. `NetTopologySuite.Geometries.ICoordinateBuffer`
1. `NetTopologySuite.Geometries.ICoordinateSequence`
1. `NetTopologySuite.Geometries.ICoordinateSequenceFactory`
1. `NetTopologySuite.Index.ISpatialIndex`
1. All interfaces in the `NetTopologySuite.IO` namespace
1. `NetTopologySuite.Operation.Buffer.IBufferParameters`

**Comment by airbreather** [comment]
## `NetTopologySuite.Operation.Buffer.IBufferParameters`

This was copied straight from GeoAPI.Core, whereas its implementation has been living in NTS.Core.

I'm guessing that it's fine to just use the class directly since that's what JTS does (it doesn't have this interface), and it looks like it was only in GeoAPI.Core to avoid having the full implementation live there, but I wanted to make sure to give an opportunity to respond before making that change.

## All interfaces in the `NetTopologySuite.IO` namespace
Broadly, these don't really seem to solve anything (**on their own**), and they have problems that I'd rather not bother fixing.  At most, they document a pattern that lots of I/O implementations follow, but even our built-in KML and GML2 implementations don't do this even though they could.

We don't use these abstractions for anything particularly interesting:
1. Our `WKBHexFileReader` uses `IBinaryGeometryReader` instead of `WKBReader` like JTS does, which means that (at best) its name and documentation are misleading.
1. There's an extension method that's almost identical to `writer?.Write(geom) ?? geom.AsBinary()`.

It's hard to imagine serious use cases that only depend on the abstractions we define here.  If, for example, you have a `Stream` or `byte[]` or `string`, and you know that it contains enough information to be able to deserialize into precisely one `Geometry`, then you almost certainly have a really good idea about what format it's in, and you might as well use an implementing class anyway.

Best-practice for text I/O is to use `TextReader` / `TextWriter`, but all text geometry readers/writers need have a `Stream` method, which means they all have to be able to deal with an encoding.

`HandleSRID` is probably often going to be pointless because not all formats will have a SRID.

`RepairRings` on the reader side might not even be relevant since the underlying format might make it outright impossible to describe a polygon with invalid rings, yet all implementations must carry this property.

The interfaces themselves are rather limiting, too.
- If you want to read / write asynchronously, then these interfaces aren't enough for you.
- If the stream specifies a collection of unrelated geometries, then you need to bundle them together as a `GeometryCollection` in order to work with this, which just feels really awkward.
    - Or, you could repeatedly call `Read` / `Write` on the same `Stream`, but then you run into the fact that some implementations infer that they have the responsibility to close the `Stream` before returning, which means the abstraction is leaky anyway.

## `NetTopologySuite.IGeometryServices`

After getting rid of other interfaces, it's really hard to imagine an implementation that's materially different from `NtsGeometryServices`.  At this point, it's pretty much a bunch of slightly abstracted constructor calls with properties to use to fill in missing method parameters.

`ReadConfiguration` / `WriteConfiguration` aren't used anywhere, and their implementations on the single implementing class don't do anything other than slow down the program.

Let's use `NtsGeometryServices` everywhere and be done with it.

## `NetTopologySuite.Geometries.ICoordinateBuffer`

The interface is only implemented in two IO modules that don't reference one another.  They can use their own implementing classes.

If we want them to share this code, then let's move `CoordinateBuffer` into NTS.Core and drop the interface.

## `NetTopologySuite.Index.ISpatialIndex`

Nobody uses this (everybody uses `NetTopologySuite.Index.ISpatialIndex<T>` instead), nobody even implements it, and it would have the same semantics as `NetTopologySuite.Index.ISpatialIndex<object>` even if anybody did.

Let's delete it.

## What remains

The others, I expect to be much more controversial (I'm still arguing with myself over them), and I'm out of time before I need to go to work today, so I won't try justifying them right away:
1. ~~`NetTopologySuite.Geometries.Prepared.IPreparedGeometry`~~ (I've changed my mind, I can't justify this to myself)
    - Instead, I've done NetTopologySuite/NetTopologySuite@5fc120b to make the existing base class, `BasicPreparedGeometry`, fit for this purpose.
    - I should also document that users should strongly favor inheriting from `BasicPreparedGeometry` over implementing `IPreparedGeometry`, as future releases may include breaking changes.
1. `NetTopologySuite.Geometries.ICoordinateSequence`
1. `NetTopologySuite.Geometries.ICoordinateSequenceFactory`
1. ~~`NetTopologySuite.Index.ISpatialIndex<T>`~~ (I've changed my mind, I can't justify this to myself)

**Comment by airbreather** [comment]
## `NetTopologySuite.Geometries.ICoordinateSequence/Factory`

This partly relies on my reasoning in NetTopologySuite/NetTopologySuite#311.  We can resolve that, without much grief on the part of implementers, by having the base classes handle making sure that everything's self-consistent, with the subclasses only being responsible for implementing the JTS-like side of the equation.

This also ties into the thing I'd mentioned above, which I will just re-quote here:

> [...] Java has default interface methods, and JTS has recently decided to start actually using them in exactly these interfaces. This feature is a long way off from being something we can use in GeoAPI*, and so using abstract classes would actually make these things easier to port, not harder, **especially** if this happens again post-v2 and we aren't comfortable with taking a breaking change.
> * *unless we drop support for the "classic" .NET Framework, which seems like a mistake and would still likely put it months away
> 
> I do understand and agree that there is an overall risk that JTS does something with one of its interfaces that we can't do because we lack multiple inheritance. I think this risk is low, but if it does happen, then it would make porting difficult.

In other words, with `ICoordinateSequence` specifically, we've already seen default interface implementations force NTS to deviate; making these abstract classes instead would actually bring NTS **closer** to JTS, in a sense, because a "default implementation" is the same as a method / property being marked `virtual`.

**Comment by airbreather** [comment]
re: this issue, I'm now satisfied with everything in the NTS project that used to be in `GeoAPI.Core`.

Apparently, I didn't include anything from `GeoAPI.CoordinateSystems` in either this or #67, so I'll close these issues now and reconsider what makes sense once #32 is done.
